### PR TITLE
Redis::OutOfMemoryError resource exception

### PR DESCRIPTION
**Pull request by fw42**
### What?

If a Redis server runs out of memory (memory usage exceeds what's defined in `maxmemory`), the server will respond to certain commands with a `Redis::CommandError`. This PR makes it so that those errors are being treated as Semian resource errors, therefore opening the Semian circuit.

### Why?

The idea is that a Redis server that is out of memory should (at least for writes) be treated similar to a Redis server that is completely down. The consequence of this PR is that the application that's using Semian sees those two errors similarly (in particular, our resiliency fallbacks in Shopify core will now catch those errors and correctly evaluate the specified fallbacks).

### What could go wrong?

One thing that's not great about this is that (I think) only **writes** would previously fail while now writes will open the circuit which will make reads fail that would previously have not failed. I'm not sure if this is a trade-off that we are willing to make? I think Semian has no distinction of reads vs. writes, so I'm not sure how to fix that. Suggestions welcome.

### Thoughts?

@es @sirupsen @casperisfine @xldenis 

(I think the test failures are unrelated to my changes, the build is failing on master too)

### CODE REVIEW

**Comment by fw42** [comment]
>  I think Semian has no distinction of reads vs. writes, so I'm not sure how to fix that.

One possibility might be to patch `Redis::Client#read` and `#write` instead of patching `#io`.

**Comment by casperisfine** [comment]
> One possibility might be to patch Redis::Client#read and #write instead of patching #io.

Unfortunately both are used for all commands, e.g. a `GET foo` will do:

```ruby
connection.write("GET foo")
connection.read
```

So I'm afraid there is no way to know the command semantic (read vs write). You could of course parse the commands, and maintain some kind of list, but with lua scripts and now plugins etc, that ain't easy.

**Comment by casperisfine** [comment]
CI seem broken in weird ways, it's probably not related to your PR.

**Comment by xldenis** [comment]
> One thing that's not great about this is that (I think) only writes would previously fail while now writes will open the circuit which will make reads fail that would previously have not failed. I'm not sure if this is a trade-off that we are willing to make? 

I'm fine with failing on both writes and reads.

**Comment by fw42** [comment]
@fbogsany what do you think about this?

**Comment by fw42** [comment]
@Shopify/caching-queuing anyone have an opinion on this?

File: lib/semian/redis.rb [hunk 1]
```diff
@@ -73,13 +80,20 @@ def resource_exceptions
...
     end
 
     def raw_semian_options
       return options[:semian] if options.key?(:semian)
       return options['semian'.freeze] if options.key?('semian'.freeze)
     end
+
+    def raise_if_out_of_memory(reply)
+      return unless reply.is_a?(::Redis::CommandError)
+      return unless reply.message == "OOM command not allowed when used memory > 'maxmemory'."
```
[the remark this answers is not in the document]
Comment (fbogsany) [reply lib/semian/redis.rb]: Just matching the `OOM` prefix will be more robust. https://redis.io/topics/protocol#resp-errors documents the format of error responses:

> The first word after the "-", up to the first space or newline, represents the kind of error returned. This is just a convention used by Redis and is not part of the RESP Error format.
>
> For example, ERR is the generic error, while WRONGTYPE is a more specific error that implies that the client tried to perform an operation against the wrong data type. This is called an Error Prefix and is a way to allow the client to understand the kind of error returned by the server without to rely on the exact message given, that may change over the time.

**Pull request merged**